File: python/remove_nth_node_from_end_of_list_19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # iterate through the list until you find the nth node from the end, then remove it.
        dummy = ListNode(0, head)
        slow = dummy
        fast = dummy
        steps = 0
        while fast.next:
            steps += 1
            if steps > n:
                slow = slow.next
            fast = fast.next
        remove = slow.next
        slow.next = remove.next

        return dummy.next

File: rust/src/remove_nth_node_from_end_of_list_19.rs
use crate::{Solution, ListNode};
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut r = ListNode { val: 0, next: head };
        let mut r = Box::new(r);
        let mut b = r.clone();
        let mut a = r.as_mut();
        let mut i = 0;

        while b.next.is_some() {
            i += 1;
            if i > n {
                a = a.next.as_mut().unwrap()
            }
            b = b.next.unwrap();
        }
        let n = a.next.as_mut().unwrap().next.take();
        a.next = n;
        return r.next;
    }
}
